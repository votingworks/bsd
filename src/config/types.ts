// Generic
export interface Dictionary<T> {
  [key: string]: T | undefined
}

// Events
export type InputEvent = React.FormEvent<EventTarget>
export type ButtonEvent = React.MouseEvent<HTMLButtonElement>
export type ButtonEventFunction = (event: ButtonEvent) => void

// Candidates
export interface Candidate {
  readonly id: string
  readonly name: string
  readonly partyId?: string
  isWriteIn?: boolean
}
export type OptionalCandidate = Candidate | undefined

// Contests
export type ContestTypes = 'candidate' | 'yesno'
export interface Contest {
  readonly id: string
  readonly districtId: string
  readonly partyId?: string
  readonly section: string
  readonly title: string
  readonly type: ContestTypes
}
export interface CandidateContest extends Contest {
  readonly type: 'candidate'
  readonly seats: number
  readonly candidates: Candidate[]
  readonly allowWriteIns: boolean
}
export interface YesNoContest extends Contest {
  readonly type: 'yesno'
  readonly description: string
  readonly shortTitle: string
}
export type Contests = (CandidateContest | YesNoContest)[]

// Election
export interface BMDConfig {
  readonly requireActivation?: boolean
  readonly showHelpPage?: boolean
  readonly showSettingsPage?: boolean
}
export interface ElectionDefaults {
  readonly bmdConfig: BMDConfig
}
export interface BallotStyle {
  readonly id: string
  readonly precincts: string[]
  readonly districts: string[]
  readonly partyId?: string
}
export interface Party {
  readonly id: string
  readonly name: string
  readonly abbrev: string
}
export type Parties = Party[]
export interface Precinct {
  readonly id: string
  readonly name: string
}
export interface District {
  readonly id: string
  readonly name: string
}
export interface Election {
  readonly ballotStyles: BallotStyle[]
  readonly parties: Parties
  readonly precincts: Precinct[]
  readonly districts: District[]
  readonly contests: Contests
  readonly county: string
  readonly date: string
  readonly seal: string
  readonly state: string
  readonly title: string
  readonly bmdConfig?: BMDConfig
}
export type OptionalElection = Election | undefined

export type SetElection = (value: OptionalElection) => void

export interface ActivationData {
  ballotStyle: BallotStyle
  precinct: Precinct
}

// Smart Card Content
export type CardDataTypes = 'voter' | 'admin'
export interface CardData {
  readonly t: CardDataTypes
}
export interface VoterCardData extends CardData {
  readonly t: 'voter'
  readonly bs: string
  readonly pr: string
}
export interface AdminCardData extends CardData {
  readonly t: 'admin'
  readonly h: string
}
