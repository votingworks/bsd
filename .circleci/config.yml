version: 2
jobs:
  build:
    docker:
      - image: circleci/python:3.7.3-node-browsers
    steps:
      - checkout
      - restore_cache: # the version on this cache changes rarely, only when a significant setting breaks the cache
          key:
            dotcache-cache-{{checksum ".circleci/config.yml" }}-{{ checksum
            "package.json" }}
      - run:
          name: Setup Dependencies
          command: yarn install --frozen-lockfile
      - run:
          name: Setup Code Climate test-reporter
          command: |
            curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./cc-test-reporter
            chmod +x ./cc-test-reporter
      - save_cache: # special step to save the dependency cache
          key:
            dotcache-cache-{{checksum ".circleci/config.yml" }}-{{ checksum
            "package.json" }}
          paths:
            - ~/.cache
      - run: # run tests
          name: Run Test and Coverage
          command: |
            # codecoverage not working yet
            #./cc-test-reporter before-build
            yarn test
            #./cc-test-reporter after-build --exit-code $?
      - run:
          name: Build static files for serving
          command: |
            yarn build
      - run:
          name: Set up python tools for backend
          command: |
            cd server
            sudo make install-smartcard
            make install-dependencies
            make install-dev-dependencies
      - run:
          name: Run Python Tests and Coverage
          command: |
            cd server
            pipenv run make coverage
