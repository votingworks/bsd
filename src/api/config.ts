import {
  GetConfigResponse,
  PatchConfigResponse,
  PatchConfigRequest,
} from '../config/types'
import fetchJSON from '../util/fetchJSON'

export async function get(): Promise<GetConfigResponse> {
  return fetchJSON<GetConfigResponse>('/config')
}

export async function patch(config: PatchConfigRequest): Promise<void> {
  const response = await fetchJSON<PatchConfigResponse>('/config', {
    method: 'PATCH',
    body: JSON.stringify(config),
    headers: { 'Content-Type': 'application/json' },
  })

  if (response.status !== 'ok') {
    throw new Error(`failed with response status: ${response.status}`)
  }
}
