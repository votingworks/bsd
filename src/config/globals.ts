import { Dictionary } from '@votingworks/ballot-encoder'

export default {
  CHECK_ICON: '✓',
  FONT_SIZES: [18, 24, 28, 32],
  TEXT_SIZE: 1,
  YES_NO_VOTES: {
    no: 'No',
    yes: 'Yes',
  },
  LANGUAGES: {
    en: 'English',
    es: 'Spanish',
  }
}

export const LANGUAGES : Dictionary<string> = {
  en: 'English',
  es: 'Spanish',
}

export const CHECK_ICON = '✓'
export const DEFAULT_LOCALE = 'en-US'
