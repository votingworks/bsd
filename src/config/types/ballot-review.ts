import * as t from '@votingworks/ballot-encoder'
import { Rect } from '@votingworks/hmpb-interpreter'
import { OkResponse } from '../types'

type ContestId = t.Contest['id']

export type GetBallotRequest = {}
export type GetBallotResponse = ReviewBallot

export interface PutBallotRequest {
  contests: MarksByContestId
}

export interface PerContestAndOption<T> {
  [key: string]: PerOption<T> | undefined
}

export interface PerOption<T> {
  [key: string]: T | undefined
}

export type PutBallotResponse = OkResponse



export interface ContestLayout {
  bounds: Rect
  options: readonly ContestOptionLayout[]
}

export interface ContestOptionLayout {
  bounds: Rect
}

export type MarksByContestId = PerContestAndOption<MarkStatus>
export type MarksByOptionId = PerOption<MarkStatus>

export enum MarkStatus {
  Marked = 'marked',
  Unmarked = 'unmarked',
  Marginal = 'marginal',
}

export type ReviewBallot =
  | ReviewMarginalMarksBallot
  | ReviewUninterpretableHmpbBallot

export interface BallotInfo {
  url: string
  image: { url: string; width: number; height: number }
}

export interface ReviewMarginalMarksBallot {
  type: 'ReviewMarginalMarksBallot'
  ballot: BallotInfo
  contests: readonly Contest[]
  layout: readonly ContestLayout[]
  marks: MarksByContestId
}

export interface ReviewUninterpretableHmpbBallot {
  type: 'ReviewUninterpretableHmpbBallot'
  ballot: BallotInfo
  contests: readonly Contest[]
}

export interface Contest {
  id: ContestId
  title: string
  bounds: Rect
  options: readonly ContestOption[]
}

export type ContestOption = CandidateContestOption | YesNoContestOption

export interface CandidateContestOption {
  type: t.CandidateContest['type']
  id: t.CandidateContest['id']
  name: t.Candidate['name']
  bounds: Rect
}

export interface YesNoContestOption {
  type: t.YesNoContest['type']
  id: Exclude<t.YesNoVote[0] | t.YesNoVote[1], undefined>
  name: string
  bounds: Rect
}
