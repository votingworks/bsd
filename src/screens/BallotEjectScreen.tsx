import React, { useState, useEffect } from 'react'
import styled from 'styled-components'

import { fetchNextBallotSheetToReview } from '../api/hmpb'
import { BallotSheetInfo } from '../config/types'

import Main, { MainChild } from '../components/Main'
import Prose from '../components/Prose'
import Button from '../components/Button'
import MainNav from '../components/MainNav'

const Columns = styled.div`
  display: flex;
  > div {
    margin-right: 1em;
    &:first-child {
      flex: 1;
    }
    &:last-child {
      margin-right: 0;
    }
    img {
      max-width: 100%;
      height: 83vh;
    }
  }
`

const RectoVerso = styled.div`
  display: flex;
  & > * {
    &:first-child {
      margin-right: 1em;
    }
  }
  img {
    max-width: 100%;
    height: 83vh;
  }
`

interface Props {
  continueScanning: () => void
}

const BallotEjectScreen = ({ continueScanning }: Props) => {
  const [sheetInfo, setSheetInfo] = useState<BallotSheetInfo | undefined>()

  useEffect(() => {
    ;(async () => {
      setSheetInfo(await fetchNextBallotSheetToReview())
    })()
  }, [setSheetInfo])

  if (!sheetInfo) {
    return null
  }

  return (
    <React.Fragment>
      <MainNav>
        <Button small onPress={continueScanning}>
          Continue Scanning Batch
        </Button>
      </MainNav>
      <Main>
        <MainChild maxWidth={false}>
          <Columns>
            <Prose maxWidth={false}>
              <h1>Remove This Ballot</h1>
              <p>
                Human review is required for this last scanned ballot.{' '}
                <strong>This ballot was not tabulated.</strong>
              </p>
              <p>
                Once this ballot has been removed, press the button to continue
                scanning.
              </p>
            </Prose>
            <RectoVerso>
              <Prose>
                <h4>Front</h4>
                <p>
                  <img src={sheetInfo.front.image.url} alt="p1" />
                </p>
              </Prose>
              <Prose>
                <h4>Back</h4>
                <p>
                  <img src={sheetInfo.back.image.url} alt="p2" />
                </p>
              </Prose>
            </RectoVerso>
          </Columns>
        </MainChild>
      </Main>
    </React.Fragment>
  )
}
export default BallotEjectScreen
