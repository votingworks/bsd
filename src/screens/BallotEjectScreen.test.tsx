import { render, waitFor, fireEvent } from '@testing-library/react'
import fetchMock from 'fetch-mock'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { BallotType, AdjudicationReason } from '@votingworks/ballot-encoder'
import { BallotSheetInfo } from '../config/types'
import BallotEjectScreen from './BallotEjectScreen'

test('says the sheet is unreadable if it is', async () => {
  const response: BallotSheetInfo = {
    id: 'mock-sheet-id',
    front: {
      image: { url: '/front/url' },
      interpretation: { type: 'BlankPage' },
    },
    back: {
      image: { url: '/back/url' },
      interpretation: { type: 'BlankPage' },
    },
  }
  fetchMock.getOnce('/scan/hmpb/review/next-sheet', response)

  const continueScanning = jest.fn()

  const { container, getByText } = render(
    <BallotEjectScreen continueScanning={continueScanning} />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  expect(container).toMatchSnapshot()

  fireEvent.click(getByText('Original Ballot Removed'))
  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith()

  continueScanning.mockClear()

  fireEvent.click(getByText('Tabulate Duplicate Ballot'))
  fireEvent.click(getByText('Tabulate Ballot and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith(true)
})

test('says the ballot sheet is overvoted if it is', async () => {
  const response: BallotSheetInfo = {
    id: 'mock-sheet-id',
    front: {
      image: { url: '/front/url' },
      interpretation: {
        type: 'InterpretedHmpbPage',
        markInfo: {
          ballotSize: { width: 1, height: 1 },
          marks: [],
        },
        metadata: {
          ballotStyleId: '1',
          precinctId: '1',
          ballotType: BallotType.Standard,
          electionHash: '',
          isTestMode: false,
          locales: { primary: 'en-US' },
          pageNumber: 1,
        },
        adjudicationInfo: {
          requiresAdjudication: true,
          allReasonInfos: [
            {
              type: AdjudicationReason.Overvote,
              contestId: '1',
              optionIds: ['1', '2'],
              expected: 1,
            },
          ],
          enabledReasons: [AdjudicationReason.Overvote],
        },
        votes: {},
      },
    },
    back: {
      image: { url: '/back/url' },
      interpretation: {
        type: 'InterpretedHmpbPage',
        markInfo: {
          ballotSize: { width: 1, height: 1 },
          marks: [],
        },
        metadata: {
          ballotStyleId: '1',
          precinctId: '1',
          ballotType: BallotType.Standard,
          electionHash: '',
          isTestMode: false,
          locales: { primary: 'en-US' },
          pageNumber: 2,
        },
        adjudicationInfo: {
          requiresAdjudication: false,
          allReasonInfos: [],
          enabledReasons: [AdjudicationReason.Overvote],
        },
        votes: {},
      },
    },
  }
  fetchMock.getOnce('/scan/hmpb/review/next-sheet', response)

  const continueScanning = jest.fn()

  const { container, getByText } = render(
    <BallotEjectScreen continueScanning={continueScanning} />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  expect(container).toMatchSnapshot()

  fireEvent.click(getByText('Original Ballot Removed'))
  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith()

  continueScanning.mockClear()

  fireEvent.click(getByText('Tabulate Duplicate Ballot'))
  fireEvent.click(getByText('Tabulate Ballot and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith(true)
})

test('says the ballot sheet is blank if it is', async () => {
  const response: BallotSheetInfo = {
    id: 'mock-sheet-id',
    front: {
      image: { url: '/front/url' },
      interpretation: {
        type: 'InterpretedHmpbPage',
        markInfo: {
          ballotSize: { width: 1, height: 1 },
          marks: [],
        },
        metadata: {
          ballotStyleId: '1',
          precinctId: '1',
          ballotType: BallotType.Standard,
          electionHash: '',
          isTestMode: false,
          locales: { primary: 'en-US' },
          pageNumber: 1,
        },
        adjudicationInfo: {
          requiresAdjudication: true,
          allReasonInfos: [{ type: AdjudicationReason.BlankBallot }],
          enabledReasons: [AdjudicationReason.Overvote],
        },
        votes: {},
      },
    },
    back: {
      image: { url: '/back/url' },
      interpretation: {
        type: 'InterpretedHmpbPage',
        markInfo: {
          ballotSize: { width: 1, height: 1 },
          marks: [],
        },
        metadata: {
          ballotStyleId: '1',
          precinctId: '1',
          ballotType: BallotType.Standard,
          electionHash: '',
          isTestMode: false,
          locales: { primary: 'en-US' },
          pageNumber: 2,
        },
        adjudicationInfo: {
          requiresAdjudication: false,
          allReasonInfos: [],
          enabledReasons: [AdjudicationReason.Overvote],
        },
        votes: {},
      },
    },
  }
  fetchMock.getOnce('/scan/hmpb/review/next-sheet', response)

  const continueScanning = jest.fn()

  const { container, getByText } = render(
    <BallotEjectScreen continueScanning={continueScanning} />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  expect(container).toMatchSnapshot()

  fireEvent.click(getByText('Original Ballot Removed'))
  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith()

  continueScanning.mockClear()

  fireEvent.click(getByText('Tabulate Duplicate Ballot'))
  fireEvent.click(getByText('Tabulate Ballot and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith(true)
})
