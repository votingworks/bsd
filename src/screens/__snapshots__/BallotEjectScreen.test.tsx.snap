// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`says the ballot sheet is blank if it is 1`] = `
<div>
  <div
    class="sc-VigVT djCdRx"
  >
    <div
      class="sc-bZQynM cWyHmc"
    >
      <div
        class="sc-gzVnrw iMtlFY"
      >
        <div
          class="sc-htoDjs fdtLZY"
        >
          <div
            class="sc-dnqmqq kKbVPv"
          >
            Voting
            <span>
              Works
            </span>
          </div>
          <div
            class="sc-iwsKbI jQhHGA"
          >
            Ballot Scanner
          </div>
        </div>
        <div
          class="sc-gZMcBi idZcZf"
        >
          <button
            class="sc-ifAKCX cAdBCb"
            disabled=""
            type="button"
          >
            Continue Scanning
          </button>
        </div>
      </div>
    </div>
    <main
      class="sc-bdVaJa kYuSve"
    >
      <div
        class="sc-bwzfXH hLOwzs"
      >
        <div
          class="sc-jTzLTM TMzWq"
        >
          <div
            class="sc-htpNat eawaaV"
          >
            <h1>
              Review Required
            </h1>
            <p>
              This last scanned sheet requires human review and
               
              <strong>
                was not tabulated
              </strong>
              .
            </p>
            <h4>
              Reason
            </h4>
            <p>
              <span>
                Reason could not be determined.
              </span>
            </p>
            <h4>
              Original Ballot Scan
            </h4>
            <p>
              Remove ballot and create a duplicate ballot for the Resolution Board to review.
              <br />
              <button
                class="sc-ifAKCX hViQmt"
                type="button"
              >
                Original Ballot Removed
              </button>
            </p>
            <h4>
              Duplicate Ballot Scan
            </h4>
            <p>
              Confirm ballot sheet was reviewed by the Resolution Board and tabulate as ballow with issue which could not be determined.
              <br />
              <button
                class="sc-ifAKCX hViQmt"
                type="button"
              >
                Tabulate Duplicate Ballot
              </button>
            </p>
          </div>
          <div
            class="sc-fjdhpX bnPSvT"
          >
            <div>
              <img
                alt="front"
                src="/front/url"
              />
            </div>
            <div>
              <img
                alt="back"
                src="/back/url"
              />
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;

exports[`says the ballot sheet is overvoted if it is 1`] = `
<div>
  <div
    class="sc-VigVT djCdRx"
  >
    <div
      class="sc-bZQynM cWyHmc"
    >
      <div
        class="sc-gzVnrw iMtlFY"
      >
        <div
          class="sc-htoDjs fdtLZY"
        >
          <div
            class="sc-dnqmqq kKbVPv"
          >
            Voting
            <span>
              Works
            </span>
          </div>
          <div
            class="sc-iwsKbI jQhHGA"
          >
            Ballot Scanner
          </div>
        </div>
        <div
          class="sc-gZMcBi idZcZf"
        >
          <button
            class="sc-ifAKCX cAdBCb"
            disabled=""
            type="button"
          >
            Continue Scanning
          </button>
        </div>
      </div>
    </div>
    <main
      class="sc-bdVaJa kYuSve"
    >
      <div
        class="sc-bwzfXH hLOwzs"
      >
        <div
          class="sc-jTzLTM TMzWq"
        >
          <div
            class="sc-htpNat eawaaV"
          >
            <h1>
              Review Required
            </h1>
            <p>
              This last scanned sheet requires human review and
               
              <strong>
                was not tabulated
              </strong>
              .
            </p>
            <h4>
              Reason
            </h4>
            <p>
              <span>
                This ballot sheet contains an 
                <strong>
                  overvote
                </strong>
                .
              </span>
            </p>
            <h4>
              Original Ballot Scan
            </h4>
            <p>
              Remove ballot and create a duplicate ballot for the Resolution Board to review.
              <br />
              <button
                class="sc-ifAKCX hViQmt"
                type="button"
              >
                Original Ballot Removed
              </button>
            </p>
            <h4>
              Duplicate Ballot Scan
            </h4>
            <p>
              Confirm ballot sheet was reviewed by the Resolution Board and tabulate as ballot sheet with an
               
              <strong>
                ovevote
              </strong>
              .
              <br />
              <button
                class="sc-ifAKCX hViQmt"
                type="button"
              >
                Tabulate Duplicate Ballot
              </button>
            </p>
          </div>
          <div
            class="sc-fjdhpX bnPSvT"
          >
            <div>
              <img
                alt="front"
                src="/front/url"
              />
            </div>
            <div>
              <img
                alt="back"
                src="/back/url"
              />
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;

exports[`says the sheet is unreadable if it is 1`] = `
<div>
  <div
    class="sc-VigVT djCdRx"
  >
    <div
      class="sc-bZQynM cWyHmc"
    >
      <div
        class="sc-gzVnrw iMtlFY"
      >
        <div
          class="sc-htoDjs fdtLZY"
        >
          <div
            class="sc-dnqmqq kKbVPv"
          >
            Voting
            <span>
              Works
            </span>
          </div>
          <div
            class="sc-iwsKbI jQhHGA"
          >
            Ballot Scanner
          </div>
        </div>
        <div
          class="sc-gZMcBi idZcZf"
        >
          <button
            class="sc-ifAKCX cAdBCb"
            disabled=""
            type="button"
          >
            Continue Scanning
          </button>
        </div>
      </div>
    </div>
    <main
      class="sc-bdVaJa kYuSve"
    >
      <div
        class="sc-bwzfXH hLOwzs"
      >
        <div
          class="sc-jTzLTM TMzWq"
        >
          <div
            class="sc-htpNat eawaaV"
          >
            <h1>
              Review Required
            </h1>
            <p>
              This last scanned sheet requires human review and
               
              <strong>
                was not tabulated
              </strong>
              .
            </p>
            <h4>
              Reason
            </h4>
            <p>
              <span>
                This sheet was 
                <strong>
                  unreadable
                </strong>
                 by the scanner.
              </span>
            </p>
            <h4>
              Original Ballot Scan
            </h4>
            <p>
              Remove ballot and create a duplicate ballot for the Resolution Board to review.
              <br />
              <button
                class="sc-ifAKCX hViQmt"
                type="button"
              >
                Original Ballot Removed
              </button>
            </p>
            <h4>
              Duplicate Ballot Scan
            </h4>
            <p>
              Confirm sheet was reviewed by the Resolution Board and tabulate as 
              <strong>
                unreadable
              </strong>
              .
              <br />
              <button
                class="sc-ifAKCX hViQmt"
                type="button"
              >
                Tabulate Duplicate Ballot
              </button>
            </p>
          </div>
          <div
            class="sc-fjdhpX bnPSvT"
          >
            <div>
              <img
                alt="front"
                src="/front/url"
              />
            </div>
            <div>
              <img
                alt="back"
                src="/back/url"
              />
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
